homeassistant/util/logging.py

```diff
@@ -164,3 +164,46 @@ def wrapper(*args: Any) -> None:
                 log_exception(*args)
         wrapper_func = wrapper
     return wrapper_func
+
+
+def catch_log_coro_exception(
+        func: Coroutine[Any, Any, Any],
+        format_err: Callable[..., Any],
+        *args: Any) -> Coroutine[Any, Any, Any]:
+    """Decorate a coroutine to catch and log exceptions."""
+    def log_exception(*args: Any) -> None:
+        module_name = inspect.getmodule(inspect.trace()[1][0]).__name__
+        # Do not print the wrapper in the traceback
+        frames = len(inspect.trace()) - 1
+        exc_msg = traceback.format_exc(-frames)
+        friendly_msg = format_err(*args)
+        logging.getLogger(module_name).error('%s\n%s', friendly_msg, exc_msg)
+
+    @wraps(func)  # type: ignore
```

Comment:
MyPy claims it's not possible to @wrap a coroutinefunction:
`error: Argument 1 to "wraps" has incompatible type "Coroutine[Any, Any, Any]"; expected "Callable[..., Any]"`

Reply:
There is a difference between a `Coroutine` and a Callable that returns a Coroutine.

Reply:
If `func` is a coroutine, then `wraps` won't work.

Reply:
`func` is a coroutine object though:
https://github.com/home-assistant/home-assistant/blob/eb21ed2986bb2326ce53875d077ab70aeaacb632/homeassistant/components/cast/media_player.py#L532-L533

Reply:
Fixed now, thanks a lot!